osquery/tables/system/windows/services.cpp

```diff
@@ -42,86 +42,108 @@ const std::map<int, std::string> kServiceType = {
     {0x00000110, "OWN_PROCESS(Interactive)"},
     {0x00000120, "SHARE_PROCESS(Interactive)"}};
 
-bool QuerySvcInfo(const SC_HANDLE& schScManager,
-                  ENUM_SERVICE_STATUS_PROCESS& svc,
-                  Row& r) {
-  DWORD cbBufSize = 0;
+auto closeServiceHandle = [](SC_HANDLE sch) {
+  if (sch != nullptr) {
+    CloseServiceHandle(sch);
+  }
+};
+typedef std::unique_ptr<SC_HANDLE__, decltype(closeServiceHandle)> svc_handle_t;
+
+static inline Status getService(const SC_HANDLE& scmHandle,
+                                const ENUM_SERVICE_STATUS_PROCESS& svc,
+                                Row& result) {
+  svc_handle_t svcHandle(
+      OpenService(scmHandle, svc.lpServiceName, SERVICE_QUERY_CONFIG),
+      closeServiceHandle);
+  if (svcHandle == nullptr) {
+    return Status(GetLastError(), "Failed to open service handle");
+  }
 
-  auto schService =
-      OpenService(schScManager, svc.lpServiceName, SERVICE_QUERY_CONFIG);
+  DWORD cbBufSize;
+  (void)QueryServiceConfig(svcHandle.get(), nullptr, 0, &cbBufSize);
+  auto err = GetLastError();
+  if (ERROR_INSUFFICIENT_BUFFER != err) {
+    return Status(err, "Failed to query size of service config buffer");
+  }
 
-  if (schService == nullptr) {
-    TLOG << "OpenService failed (" << GetLastError() << ")";
-    return FALSE;
+  std::unique_ptr<QUERY_SERVICE_CONFIG> lpsc(
+      static_cast<LPQUERY_SERVICE_CONFIG>(malloc(cbBufSize)));
+  if (lpsc == nullptr) {
+    return Status(1, "Failed to malloc service config buffer");
   }
 
-  QueryServiceConfig(schService, nullptr, 0, &cbBufSize);
-  auto lpsc = (LPQUERY_SERVICE_CONFIG)malloc(cbBufSize);
-  if (!QueryServiceConfig(schService, lpsc, cbBufSize, &cbBufSize)) {
-    TLOG << "QueryServiceConfig failed (" << GetLastError() << ")";
+  if (0 ==
+      QueryServiceConfig(svcHandle.get(), lpsc.get(), cbBufSize, &cbBufSize)) {
```

Comment:
ERROR_SUCCESS would be the result from GetLastError() vs the return code, but I see your point, the code is ugly. The MS examples actually recommend to just directly check the result (https://msdn.microsoft.com/en-us/library/windows/desktop/ms684928(v=vs.85).aspx), but that seems to be somewhat against the code style we use, so i'm fine to change it to store&check.